synapse-module/nginx.nix

```diff
@@ -198,7 +198,8 @@ in
     };
 
     services.nginx.virtualHosts."${cfg.public_baseurl}" = {
-      enableACME = true;
+      enableACME = cfg.enableACME;
+      useACMEHost = cfg.useACMEHost;
```

Comment:
An alternative to doing this is to just do

```
enableAcme = lib.mkDefault true;
```

And then if you want to override it you can configure nginx yourself:

```
services.nginx."${config.services.matrix-synapse-next.public_baseurl}" = {
  enableACME = false;
  useACMEHost = "foobar";
};
```

Which I think I generally prefer over exposing internal nginx options here

Reply:
(also please note that `useACMEHost` requires `enableACME` to be set to true)

Reply:
Heya.

> `useACMEHost` requires `enableACME` to be set to true

This is not correct. In fact, [making sure not both are enabled is enforced in the nginx module](https://github.com/NixOS/nixpkgs/blob/96f7ceb7b980489b79e9cc084bdb4b9481d421a6/nixos/modules/services/web-servers/nginx/default.nix#L1128-L1134)

> but I don't know how is the best practice to enforce that in nix

But if it was correct, you could enforce it by setting the others default to be true if this one was true, and enforce it with [an assertion](https://nixos.org/manual/nixos/stable/index.html#sec-assertions), just like the one in the nginx module.

Reply:
`enableACME` will request a cert, it's not needed (and probably should not be used) if you use `useACMEHost`, since you'll be generating one through [`security.acme.certs`](https://search.nixos.org/options?channel=unstable&show=security.acme.certs&query=security.acme.certs)

Some modules expose their virtualhost configuration in the module as an option that is then merged with the default values. [see this search for examples](https://search.nixos.org/options?channel=23.05&from=0&size=50&sort=relevance&type=packages&query=nginx). But I think this is really unnecessary. nginx isn't the only thing one has to patch like this in a nixos system. Sometimes you might need to add extra configuration to the underlying systemd unit files - like an after or requires. Sometimes you might need to edit the service users joined groups. I don't think its reasonable to add the whole systemd config options, the entire nginx vhost option set, or the users options in the module.

Now that's not really an argument against extracting _single_ options from those sets, like you're doing  here. But that's the main line of reasoning against lifting up the abstractions.

Reply:
Thank you! I understand better now.